### PR TITLE
[AMDGPU] Fix mul combine for MUL24

### PR DESCRIPTION
**Pull request by Pierre-vh**
MUL24 can now return  a i64 for i32 operands, but the combine was never updated to handle this case. Extend the operand when rewriting the ADD to handle it.

Fixes SWDEV-436654

### CODE REVIEW

**Review by jayfoad — COMMENTED** [review]
"mul24 x, (add y, 1) -> add (mul24 x, y), x" does not seem safe at all. You know that y+1 is a 24-bit value, but that does not guarantee that y is a 24-bit value.

**Comment by Pierre-vh** [comment]
> "mul24 x, (add y, 1) -> add (mul24 x, y), x" does not seem safe at all. You know that y+1 is a 24-bit value, but that does not guarantee that y is a 24-bit value.

Ah indeed, good catch.
Should the combine just be removed? (cc @arsenm - I think you wrote this one)

**Comment by jayfoad** [comment]
> > "mul24 x, (add y, 1) -> add (mul24 x, y), x" does not seem safe at all. You know that y+1 is a 24-bit value, but that does not guarantee that y is a 24-bit value.
> 
> Ah indeed, good catch. Should the combine just be removed?

I guess it's safe for a regular mul, so maybe move the "Skip if already mul24" up a bit?

**Comment by arsenm** [comment]
This probably needs work to extend for the mul24 cases, with some known bits checks. Probably should just skip for it

File: llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp [hunk 1]
```diff
@@ -4254,10 +4254,6 @@ SDValue AMDGPUTargetLowering::performMulCombine(SDNode *N,
     return DAG.getNode(ISD::ADD, DL, VT, MulVal, N0);
   }
 
-  // Skip if already mul24.
-  if (N->getOpcode() != ISD::MUL)
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp:4258]: Probably should turn this into an assert? Is there definitely nothing else in performMulCombine that can operate on mul24?

File: llvm/test/CodeGen/AMDGPU/mul-combine-crash.ll [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+; CHECK-NEXT:    v_mov_b32_e32 v1, 0
+; CHECK-NEXT:    s_setpc_b64 s[30:31]
+entry:
+  %a = add i64 %x, 0
+  %b = and i64 %a, 255
+  %c = and i32 %z, 1
+  %d = add nuw nsw i32 %c, 1
+  %e = zext nneg i32 %d to i64
+  %f = mul nuw nsw i64 %e, %b
+  %g = add nuw nsw i64 %f, 0
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/mul-combine-crash.ll:43]: Are the adds of 0 really necessary? Can the test merge in with the existing test?

**Review by arsenm — APPROVED** [review]

**Pull request merged**